File: artifacts/tulipindicators/translate/src/hma.rs
use crate::*;
use libm::sqrt;


pub fn ti_hma_start(options: Option<&[f64]>) -> Option<i32> {
    // Check if options is None or empty (equivalent to NULL check in C)
    let options = options?;
    if options.is_empty() {
        return None;
    }

    let period = options[0] as i32;
    let periodsqrt = sqrt(options[0]) as i32;
    Some((period + periodsqrt) - 2)
}
pub fn helper_ti_hma_1(
    output_idx_ref: &mut u32,
    sum_ref: &mut f64,
    weight_sum_ref: &mut f64,
    sum2_ref: &mut f64,
    weight_sum2_ref: &mut f64,
    sumsqrt_ref: &mut f64,
    weight_sumsqrt_ref: &mut f64,
    input: &[f64],
    period: i32,
    output: &mut [f64],
    period2: i32,
    periodsqrt: i32,
    weights: f64,
    weights2: f64,
    weightssqrt: f64,
    i: i32,
    buff: &mut TiBuffer,
) {
    let mut output_idx = *output_idx_ref;
    let mut sum = *sum_ref;
    let mut weight_sum = *weight_sum_ref;
    let mut sum2 = *sum2_ref;
    let mut weight_sum2 = *weight_sum2_ref;
    let mut sumsqrt = *sumsqrt_ref;
    let mut weight_sumsqrt = *weight_sumsqrt_ref;

    let i_usize = i as usize;
    weight_sum += input[i_usize] * period as f64;
    sum += input[i_usize];
    weight_sum2 += input[i_usize] * period2 as f64;
    sum2 += input[i_usize];

    let wma = weight_sum / weights;
    let wma2 = weight_sum2 / weights2;
    let diff = (2.0 * wma2) - wma;

    weight_sumsqrt += diff * periodsqrt as f64;
    sumsqrt += diff;

    {
        buff.vals[buff.index as usize] = diff;
        buff.index += 1;
        if buff.index >= buff.size {
            buff.index = 0;
        }
    }

    if i >= ((period - 1) + (periodsqrt - 1)) {
        output[output_idx as usize] = weight_sumsqrt / weightssqrt;
        output_idx += 1;
        weight_sumsqrt -= sumsqrt;
        let buff_size = buff.size as usize;
        let index = ((buff.index + buff.size - 1) + 1) as usize % buff_size;
        sumsqrt -= buff.vals[index];
    } else {
        weight_sumsqrt -= sumsqrt;
    }

    weight_sum -= sum;
    sum -= input[(i - period + 1) as usize];
    weight_sum2 -= sum2;
    sum2 -= input[(i - period2 + 1) as usize];

    *output_idx_ref = output_idx;
    *sum_ref = sum;
    *weight_sum_ref = weight_sum;
    *sum2_ref = sum2;
    *weight_sum2_ref = weight_sum2;
    *sumsqrt_ref = sumsqrt;
    *weight_sumsqrt_ref = weight_sumsqrt;
}

pub fn ti_hma(
    size: i32,
    inputs: Option<&[&[f64]]>,
    options: Option<&[f64]>,
    outputs: Option<&mut [&mut [f64]]>,
) -> Option<i32> {
    // Check for null inputs/outputs
    if inputs.is_none() || options.is_none() || outputs.is_none() {
        return Some(1);
    }

    let inputs = inputs.unwrap();
    let options = options.unwrap();
    let outputs = outputs.unwrap();

    let input = inputs[0];
    let period = options[0] as i32;
    let output = &mut outputs[0];

    if period < 1 {
        return Some(1);
    }

    let start = match ti_hma_start(Some(options)) {
        Some(s) => s,
        None => return Some(1),
    };

    if size <= start {
        return Some(0);
    }

    let period2 = (period / 2) as i32;
    let periodsqrt = sqrt(period as f64) as i32;
    let weights = (period * (period + 1)) as f64 / 2.0;
    let weights2 = (period2 * (period2 + 1)) as f64 / 2.0;
    let weightssqrt = (periodsqrt * (periodsqrt + 1)) as f64 / 2.0;

    let mut sum = 0.0;
    let mut weight_sum = 0.0;
    let mut sum2 = 0.0;
    let mut weight_sum2 = 0.0;
    let mut sumsqrt = 0.0;
    let mut weight_sumsqrt = 0.0;
    let mut output_idx = 0;

    for i in 0..(period - 1) as usize {
        weight_sum += input[i] * (i + 1) as f64;
        sum += input[i];

        if i >= (period - period2) as usize {
            weight_sum2 += input[i] * ((i + 1) - (period - period2) as usize) as f64;
            sum2 += input[i];
        }
    }

    let mut buff = match ti_buffer_new(periodsqrt) {
        Some(b) => b,
        None => return Some(1),
    };

    for i in (period - 1)..size {
        helper_ti_hma_1(
            &mut output_idx,
            &mut sum,
            &mut weight_sum,
            &mut sum2,
            &mut weight_sum2,
            &mut sumsqrt,
            &mut weight_sumsqrt,
            input,
            period,
            output,
            period2,
            periodsqrt,
            weights,
            weights2,
            weightssqrt,
            i,
            &mut buff,
        );
    }

    ti_buffer_free(*buff);

    assert!((output[output_idx as usize] - outputs[0][0]) == (size - start) as f64);
    Some(0)
}

File: artifacts/tulipindicators/decompose/src/hma.c
int ti_hma_start(const double *options)
{
  const int period = (int) options[0];
  const int periodsqrt = (int) sqrt(period);
  return (period + periodsqrt) - 2;
}


/*** DEPENDENCIES:
***/


// hint:  ['output_idx_ref is a mutable refrence to unsigned int', 'sum_ref is a mutable refrence to double', 'weight_sum_ref is a mutable refrence to double', 'sum2_ref is a mutable refrence to double', 'weight_sum2_ref is a mutable refrence to double', 'sumsqrt_ref is a mutable refrence to double', 'weight_sumsqrt_ref is a mutable refrence to double']
void helper_ti_hma_1(unsigned int * const output_idx_ref, double * const sum_ref, double * const weight_sum_ref, double * const sum2_ref, double * const weight_sum2_ref, double * const sumsqrt_ref, double * const weight_sumsqrt_ref, const double * const input, const int period, double * const output, const int period2, const int periodsqrt, const double weights, const double weights2, const double weightssqrt, int i, ti_buffer * const buff)
{
  unsigned int output_idx = *output_idx_ref;
  double sum = *sum_ref;
  double weight_sum = *weight_sum_ref;
  double sum2 = *sum2_ref;
  double weight_sum2 = *weight_sum2_ref;
  double sumsqrt = *sumsqrt_ref;
  double weight_sumsqrt = *weight_sumsqrt_ref;
  weight_sum += input[i] * period;
  sum += input[i];
  weight_sum2 += input[i] * period2;
  sum2 += input[i];
  const double wma = weight_sum / weights;
  const double wma2 = weight_sum2 / weights2;
  const double diff = (2 * wma2) - wma;
  weight_sumsqrt += diff * periodsqrt;
  sumsqrt += diff;
  do
  {
    buff->vals[buff->index] = diff;
    buff->index = buff->index + 1;
    if (buff->index >= buff->size)
    {
      buff->index = 0;
    }
  }
  while (0);
  if (i >= ((period - 1) + (periodsqrt - 1)))
  {
    output[output_idx] = weight_sumsqrt / weightssqrt;
    output_idx += 1;
    weight_sumsqrt -= sumsqrt;
    sumsqrt -= buff->vals[(((buff->index + buff->size) - 1) + 1) % buff->size];
  }
  else
  {
    weight_sumsqrt -= sumsqrt;
  }
  weight_sum -= sum;
  sum -= input[(i - period) + 1];
  weight_sum2 -= sum2;
  sum2 -= input[(i - period2) + 1];
  *output_idx_ref = output_idx;
  *sum_ref = sum;
  *weight_sum_ref = weight_sum;
  *sum2_ref = sum2;
  *weight_sum2_ref = weight_sum2;
  *sumsqrt_ref = sumsqrt;
  *weight_sumsqrt_ref = weight_sumsqrt;
}


/*** DEPENDENCIES:
typedef struct 
{
  int size;
  int pushes;
  int index;
  double sum;
  double vals[1];
} ti_buffer
----------------------------
***/


int ti_hma(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *input = inputs[0];
  const int period = (int) options[0];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_hma_start(options))
  {
    return 0;
  }
  const int period2 = (int) (period / 2);
  const int periodsqrt = (int) sqrt(period);
  const double weights = (period * (period + 1)) / 2;
  const double weights2 = (period2 * (period2 + 1)) / 2;
  const double weightssqrt = (periodsqrt * (periodsqrt + 1)) / 2;
  double sum = 0;
  double weight_sum = 0;
  double sum2 = 0;
  double weight_sum2 = 0;
  double sumsqrt = 0;
  double weight_sumsqrt = 0;
  int i;
  for (i = 0; i < (period - 1); i += 1)
  {
    weight_sum += input[i] * (i + 1);
    sum += input[i];
    if (i >= (period - period2))
    {
      weight_sum2 += input[i] * ((i + 1) - (period - period2));
      sum2 += input[i];
    }
  }

  ti_buffer *buff = ti_buffer_new(periodsqrt);
  for (i = period - 1; i < size; i += 1)
  {
    helper_ti_hma_1(&output_idx, &sum, &weight_sum, &sum2, &weight_sum2, &sumsqrt, &weight_sumsqrt, input, period, output, period2, periodsqrt, weights, weights2, weightssqrt, i, buff);
  }

  ti_buffer_free(buff);
  assert(((&output[output_idx]) - outputs[0]) == (size - ti_hma_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_hma_start(const double *options)
{
  const int period = (int) options[0];
  const int periodsqrt = (int) sqrt(period);
  return (period + periodsqrt) - 2;
}


----------------------------
void ti_buffer_free(ti_buffer *buffer)
{
  free(buffer);
}


----------------------------
ti_buffer *ti_buffer_new(int size)
{
  const int s = ((int) (sizeof(ti_buffer))) + ((size - 1) * ((int) (sizeof(double))));
  ti_buffer *ret = (ti_buffer *) malloc((unsigned int) s);
  ret->size = size;
  ret->pushes = 0;
  ret->index = 0;
  ret->sum = 0;
  return ret;
}


----------------------------
void helper_ti_hma_1(unsigned int * const output_idx_ref, double * const sum_ref, double * const weight_sum_ref, double * const sum2_ref, double * const weight_sum2_ref, double * const sumsqrt_ref, double * const weight_sumsqrt_ref, const double * const input, const int period, double * const output, const int period2, const int periodsqrt, const double weights, const double weights2, const double weightssqrt, int i, ti_buffer * const buff)
{
  unsigned int output_idx = *output_idx_ref;
  double sum = *sum_ref;
  double weight_sum = *weight_sum_ref;
  double sum2 = *sum2_ref;
  double weight_sum2 = *weight_sum2_ref;
  double sumsqrt = *sumsqrt_ref;
  double weight_sumsqrt = *weight_sumsqrt_ref;
  weight_sum += input[i] * period;
  sum += input[i];
  weight_sum2 += input[i] * period2;
  sum2 += input[i];
  const double wma = weight_sum / weights;
  const double wma2 = weight_sum2 / weights2;
  const double diff = (2 * wma2) - wma;
  weight_sumsqrt += diff * periodsqrt;
  sumsqrt += diff;
  do
  {
    buff->vals[buff->index] = diff;
    buff->index = buff->index + 1;
    if (buff->index >= buff->size)
    {
      buff->index = 0;
    }
  }
  while (0);
  if (i >= ((period - 1) + (periodsqrt - 1)))
  {
    output[output_idx] = weight_sumsqrt / weightssqrt;
    output_idx += 1;
    weight_sumsqrt -= sumsqrt;
    sumsqrt -= buff->vals[(((buff->index + buff->size) - 1) + 1) % buff->size];
  }
  else
  {
    weight_sumsqrt -= sumsqrt;
  }
  weight_sum -= sum;
  sum -= input[(i - period) + 1];
  weight_sum2 -= sum2;
  sum2 -= input[(i - period2) + 1];
  *output_idx_ref = output_idx;
  *sum_ref = sum;
  *weight_sum_ref = weight_sum;
  *sum2_ref = sum2;
  *weight_sum2_ref = weight_sum2;
  *sumsqrt_ref = sumsqrt;
  *weight_sumsqrt_ref = weight_sumsqrt;
}


----------------------------
typedef struct 
{
  int size;
  int pushes;
  int index;
  double sum;
  double vals[1];
} ti_buffer
----------------------------
***/


